lib/src/importer/canonicalize_context.dart

```diff
@@ -0,0 +1,49 @@
+// Copyright 2014 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:meta/meta.dart';
+
+/// Contextual information used by importers' `canonicalize` method.
+@internal
+final class CanonicalizeContext {
+  /// Whether the Sass compiler is currently evaluating an `@import` rule.
+  bool get fromImport => _fromImport;
+  bool _fromImport;
+
+  /// The URL of the stylesheet that contains the current load.
+  Uri? get containingUrl {
+    _wasContainingUrlAccessed = true;
+    return _containingUrl;
+  }
+
+  final Uri? _containingUrl;
+
+  /// Returns the same value as [containingUrl], but doesn't mark it accessed.
+  Uri? get containingUrlWithoutMarking => _containingUrl;
+
+  /// Whether [containingUrl] has been accessed.
+  ///
+  /// This is used to determine whether canonicalize result is cacheable.
+  ///
+  /// @nodoc
+  @internal
+  bool get wasContainingUrlAccessed => _wasContainingUrlAccessed;
+  var _wasContainingUrlAccessed = false;
+
+  /// Runs [callback] in a context with specificed [fromImport].
+  ///
+  /// @nodoc
+  @internal
+  T withFromImport<T>(bool fromImport, T callback()) {
```

Comment:
Due to the access tracking of containingUrl, `CanonicalizeContext` by nature is already mutable, and if we fork another copy to modify import, it will make the access tracking complicated as the fork need to propagate the access tracking to the original copy.

In the current implementation, `CanonicalizeContext` is only initialized as zone variable, and `withFromImport` is only called on within zones, so that the mutation should be safe in practice. I understand there is risk if it is called un-zoned, so I added an assertion to make this method is only callable when the CanonicalizeContext being invoked is the current context inside the current zone.